src/core/utilities/box/hollow_box.h

```diff
@@ -0,0 +1,225 @@
+
+#ifndef PHARE_CORE_UTILITIES_BOX_HOLLOW_BOX_H
+#define PHARE_CORE_UTILITIES_BOX_HOLLOW_BOX_H
+
+#include <cstddef>
+
+#include "core/data/grid/gridlayout.h"
+#include "core/data/ndarray/ndarray_vector.h"
+#include "core/utilities/point/point.h"
+#include "core/utilities/meta/meta_utilities.h"
+#include "core/utilities/types.h"
+
+
+namespace PHARE::core
+{
+template<std::size_t dim_, typename Size = std::uint32_t>
+class HollowBox
+{
+public:
+    static constexpr auto dim = dim_;
+
+    HollowBox(std::array<Size, dim> const& sizes, std::array<Size, dim> const& buffer)
+        : sizes_{sizes}
+        , buffer_{buffer}
+    {
+    }
+
+    std::size_t cells() const
+    {
+        if constexpr (dim == 1)
+            return ((buffer_[0]) * 2);
+
+        if constexpr (dim == 2)
+            return (sizes_[1] * ((buffer_[0]) * 2)) + (sizes_[0] * ((buffer_[1]) * 2))
+                   - ((buffer_[0]) * (buffer_[1]) * 4);
+
+        throw std::runtime_error("not implemented");
+    }
+
+    template<typename T, typename V>
+    void setBuffer(T const* data, V val)
+    {
+        core::NdArrayView<dim, T> view{data, sizes_};
+
+        if constexpr (dim == 1)
+        {
+            assert(4 == buffer_[0]);
+            for (std::size_t i = 0; i < buffer_[0]; i++)
+                view(i) = val;
+            assert(4 == sizes_[0] - (sizes_[0] - buffer_[0]));
+            for (std::size_t i = sizes_[0] - buffer_[0]; i < sizes_[0]; i++)
+                view(i) = val;
+        }
+        else if constexpr (dim == 2)
+        {
+            for (std::size_t i = 0; i < buffer_[0]; i++) // bottom
+                for (std::size_t j = 0; j < sizes_[1]; j++)
+                    view(i, j) = val;
+
+            for (std::size_t i = sizes_[0] - buffer_[0]; i < sizes_[0]; i++) // top
+                for (std::size_t j = 0; j < sizes_[1]; j++)
+                    view(i, j) = val;
+
+            for (std::size_t i = buffer_[0]; i < sizes_[0] - buffer_[0]; i++)
+            {
+                for (std::size_t j = 0; j < buffer_[1]; j++)
+                    view(i, j) = val; // left
+
+                for (std::size_t j = sizes_[1] - buffer_[1]; j < sizes_[1]; j++)
+                    view(i, j) = val; // left
+            }
+        }
+        else if constexpr (dim == 3)
+        {
+            throw std::runtime_error("not implemented");
+        }
+    }
+
+private:
+    std::array<Size, dim> sizes_;
+    std::array<Size, dim> buffer_;
+};
+
+template<typename GridLayout>
+auto totalCells(GridLayout const& layout, QtyCentering centering)
```

Comment:
i don't think this function is named after what it does.
Also, you may want to look at https://github.com/PHAREHUB/PHARE/blob/7d11552dc86f6a8fbb963146eef498583658801a/src/core/data/grid/gridlayout.h#L850

Reply:
done